### PR TITLE
Feature/fetch delete todos

### PR DESCRIPTION
**Pull request by ghost**
fetch todos from endpoints
show loading message
show no task on empty todo list
delete todo from endpoint and display remainings

### CODE REVIEW

File: frontend/src/App.css [hunk 1]
```diff
@@ -37,6 +37,12 @@ body {
   list-style-type: none;
 }
 
+.h3-msg {
```
Comment (ZunairaT) [inline frontend/src/App.css]: `.msg-text` would be a better name. so when it is reused with other selectors it doesnt feel odd

File: frontend/src/App.tsx [hunk 2]
```diff
@@ -1,17 +1,35 @@
...
+
 function App() {
   const [todos, setTodos] = useState<Todo[]>([]);
   const [newTodoDescription, setNewTodoDescription] = useState("");
+  const [loading, setLoading] = useState(true);
 
-  const onChange = (name: string, value: string) => {
-    setNewTodoDescription(value);
-  };
+  async function fetchTodos(url: string) {
```
Comment (alinauman02) [inline frontend/src/App.tsx]: Remove `async` from this function, as we are not using `await` inside it.

File: frontend/src/App.tsx [hunk 3]
```diff
@@ -45,6 +62,27 @@ function App() {
     setNewTodoDescription("");
   };
 
+  async function delTodo(url: string, id: string) {
```
Comment (alinauman02) [inline frontend/src/App.tsx]: Remove `async` from this function, as we are not using `await` inside it.

File: frontend/src/App.tsx [hunk 4]
```diff
@@ -60,8 +98,13 @@ function App() {
             Add
           </button>
         </div>
-
-        <TodoList todos={todos} />
+        {loading && <h3 className="h3-msg">LOADING...</h3>}
```
Comment (alinauman02) [inline frontend/src/App.tsx]: Create a regular `const` variable and assign a value to it based on these conditions, and use that variable to render the content inside the component template.

File: frontend/src/App.tsx [hunk 5]
```diff
@@ -45,6 +62,27 @@ function App() {
...
+  async function delTodo(url: string, id: string) {
+    return fetch(url + id, {
+      method: "DELETE",
+      headers: {
+        "Content-Type": "application/json",
+      },
+    });
+  }
+
+  const deleteTodo = async (id: string) => {
```
Comment (alinauman02) [inline frontend/src/App.tsx]: Rename this to `onDeleteTodo`

**Review by alinauman02 — CHANGES_REQUESTED** [review]

**Comment by alinauman02** [comment]
Also rename `addTask` function to `onAddTodo`

File: frontend/src/App.tsx (hunk 5 above)
Comment (ZunairaT) [inline frontend/src/App.tsx]: deleteTodo and delTodo sound very similar and hence confusing. you can name the function that is called first on user click `onDeleteTodo` and the other one `deleteTodo`

**Review by ZunairaT — CHANGES_REQUESTED** [review]

File: frontend/src/App.tsx [hunk 6]
```diff
@@ -1,19 +1,47 @@
...
+
 function App() {
   const [todos, setTodos] = useState<Todo[]>([]);
   const [newTodoDescription, setNewTodoDescription] = useState("");
+  const [loading, setLoading] = useState(true);
 
-  const onChange = (name: string, value: string) => {
-    setNewTodoDescription(value);
-  };
+  function fetchTodos(url: string, signal: AbortSignal) {
```
Comment (alinauman02) [inline frontend/src/App.tsx]: You are using a mixture of regular functions and arrow functions. Using either is fine, but you have to be consistent. Either declare all functions with the `function` keyword, or make them all arrow functions.

File: frontend/src/App.tsx [hunk 7]
```diff
@@ -1,19 +1,47 @@
...
+  }
+
+  useEffect(() => {
+    const controller = new AbortController();
+    const loadTodos = async () => {
+      try {
+        const response = await fetchTodos(urlString, controller.signal);
+        const tempTodos: Todo[] = await response.json();
+        setTodos(tempTodos);
+        setLoading(false);
```
Comment (alinauman02) [inline frontend/src/App.tsx:32]: There is a problem here. If the request fails for some reason, `loading` will not be set to `false`, and the user will always see the loading state.

Fix this such that `loading` is set to `false` in both the cases.

File: frontend/src/App.tsx [hunk 8]
```diff
@@ -45,6 +72,36 @@ function App() {
...
+    setNewTodoDescription(value);
+  };
+
+  const content = loading ? (
+    <h3 className="msg-text">LOADING...</h3>
+  ) : !loading && todos.length !== 0 ? (
+    <TodoList todos={todos} deleteTodo={onDeleteTodo}></TodoList>
+  ) : (
+    !loading &&
+    todos.length === 0 && <h3 className="msg-text">NO TASKS ADDED</h3>
```
Comment (alinauman02) [inline frontend/src/App.tsx]: I mentioned this before too, but we are calling them Todos, not Tasks. Change this to "No Todos"

File: frontend/src/App.tsx [hunk 9]
```diff
@@ -45,6 +72,36 @@ function App() {
...
+  );
+
   return (
     <div className="todo-app-wrapper">
       <div className=" card">
```
Comment (alinauman02) [inline frontend/src/App.tsx:108]: You should make a change here. Any `input` field should ideally be inside a `form` element. This is semantically more appropriate.

A `form` also needs a `submit` button. You already have one here, and you can tweak that to submit this form.

If done correctly, you can submit the form (that is, "click" the Add button) by hitting the Enter key. In other words, you can create a Todo by just entering the description and hitting the Enter key.

**Note: Submitting the form should not refresh the page.**

**Review by alinauman02 — CHANGES_REQUESTED** [review]
Added some new comments.

File: frontend/src/App.tsx [hunk 10]
```diff
@@ -52,7 +52,8 @@ function App() {
     });
   };
 
-  const onAddTodo = async () => {
+  const onAddTodo = async (event: FormEvent<HTMLFormElement>) => {
+    event.preventDefault();
```
Comment (alinauman02) [inline frontend/src/App.tsx:56]: Nice! :+1:

**Review by alinauman02 — APPROVED** [review]

**Review by ZunairaT — APPROVED** [review]

**Pull request merged**